src/util/git-ref-resolver.js

```diff
@@ -0,0 +1,192 @@
+/* @flow */
+
+import type Config from '../config.js';
+import {removeSuffix} from './misc.js';
+
+const semver = require('semver');
+
+export type GitRefs = {
+  [name: string]: string,
+};
+export type DefaultBranch = {defaultBranch: true};
+export type ResolvedSha = {sha: string, ref: ?string};
+type Names = {tags: Array<string>, branches: Array<string>};
+
+export const isCommitSha = (target: string): boolean => Boolean(target) && /^[a-f0-9]{5,40}$/.test(target);
+
+const REF_TAG_PREFIX = 'refs/tags/';
+const REF_BRANCH_PREFIX = 'refs/heads/';
+
+// This regex is designed to match output from git of the style:
+//   ebeb6eafceb61dd08441ffe086c77eb472842494  refs/tags/v0.21.0
+// and extract the hash and ref name as capture groups
+const gitRefLineRegex = /^([a-fA-F0-9]+)\s+(refs\/(?:tags|heads)\/.*)$/;
+
+const refNameRegexp = /^refs\/(tags|heads)\/(.+)$/;
+
+class GitRefResolver {
+  constructor(config: Config, version: string, refs: GitRefs) {
+    this.config = config;
+    this.version = version;
+    this.refs = refs;
+  }
+
+  config: Config;
+  version: string;
+  refs: GitRefs;
+  names: ?Names;
+
+  async resolve(): Promise<DefaultBranch | ResolvedSha | false> {
+    const testFunctions = [
+      () => this.tryVersionAsGitCommit(),
+      () => (this.version === '' ? this.useDefaultBranch() : false),
+      () => this.tryVersionAsFullRef(),
+      () => this.tryVersionAsTagName(),
+      () => this.tryVersionAsBranchName(),
+      () => this.tryVersionAsTagSemver(),
+      () => this.tryVersionAsBranchSemver(),
+      () => (this.version === '*' ? this.useDefaultBranch() : false),
+    ];
+    for (const testFunction of testFunctions) {
+      const result = await testFunction();
+      if (result !== false) {
+        return result;
+      }
+    }
+    return false;
+  }
+
+  useDefaultBranch(): DefaultBranch {
+    return {defaultBranch: true};
+  }
+
+  tryVersionAsGitCommit(): ResolvedSha | false {
+    if (isCommitSha(this.version)) {
+      for (const ref in this.refs) {
+        const sha = this.refs[ref];
+        if (sha.startsWith(this.version)) {
+          return {sha, ref};
+        }
+      }
+      return {sha: this.version, ref: undefined};
+    }
+    return false;
+  }
+
+  tryRef(ref: string): ResolvedSha | false {
+    if (this.refs[ref]) {
+      return {
+        sha: this.refs[ref],
+        ref,
+      };
+    }
+    return false;
+  }
+
+  tryVersionAsFullRef(): ResolvedSha | false {
+    if (this.version.startsWith('refs/')) {
+      return this.tryRef(this.version);
+    }
+    return false;
+  }
+
+  tryVersionAsTagName(): ResolvedSha | false {
+    const ref = `${REF_TAG_PREFIX}${this.version}`;
+    return this.tryRef(ref);
+  }
+
+  tryVersionAsBranchName(): ResolvedSha | false {
+    const ref = `${REF_BRANCH_PREFIX}${this.version}`;
+    return this.tryRef(ref);
+  }
+
+  getSemverNames(): Names {
+    if (!this.names) {
```

Comment:
That's a preference only for getters, most of the time I also prefer putting an if/return at the beginning and avoid indentation.
In the case of the getters:
- A single `return` statement
- Closer to the implementation of a simple getter
  ```
  getValue() {
    // if you add/remove lazy-init later, just change this line
    return this.value;
  }
  ```
- Ensure that the "value" is saved, to avoid errors like:
  ```
  getValue() {
    if (this.value) {
      return this.value;
    }
    const value = computeValue();
    // ...
    return value;
  }
  ```